From 605b05c884baa1fcb1c1d9447949e4c0a47197bf Mon Sep 17 00:00:00 2001
From: Ramalingam C <ramalingam.c@intel.com>
Date: Mon, 12 Oct 2015 17:24:12 +0530
Subject: [PATCH] FOR_UPSTREAM [VPG]: drm/i915/cht: Dont wait for PIPE OFF

In case of CHT DSI command mode configuration, when we disable
the PIPE trough PIPE_CONF register, PIPE state is not getting
updated.

So for the above scenario this patch helps us to skip the wait
for PIPE OFF base on pipe state bit.

Change-Id: I8cd002f06c9f62e79359a17ed43ac2526e0b3574
Signed-off-by: Ramalingam C <ramalingam.c@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-9520
(cherry picked from commit 2a410811906952c4d14b27fc028e2893b7fa324b)
---
 drivers/gpu/drm/i915/intel_display.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index bcbdaad..2f8e1c1 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -2310,6 +2310,7 @@ static void intel_disable_pipe(struct drm_i915_private *dev_priv,
 	struct drm_crtc *crtc = dev_priv->pipe_to_crtc_mapping[pipe];
 	struct drm_device *dev = crtc->dev;
 	int reg;
+	bool cht_dsi_cmd_mode = false;
 	u32 val;
 
 	/*
@@ -2332,14 +2333,19 @@ static void intel_disable_pipe(struct drm_i915_private *dev_priv,
 	for_each_encoder_on_crtc(dev, crtc, encoder) {
 		if (encoder->type == INTEL_OUTPUT_DSI) {
 			intel_dsi = enc_to_intel_dsi(&encoder->base);
-			if (intel_dsi && is_cmd_mode(intel_dsi))
+			if (intel_dsi && is_cmd_mode(intel_dsi)) {
+				cht_dsi_cmd_mode = IS_CHERRYVIEW(dev_priv->dev);
 				val = val & ~PIPECONF_MIPI_DSR_ENABLE;
+			}
 			break;
 		}
 	}
 
 	I915_WRITE(reg, val & ~PIPECONF_ENABLE);
-	intel_wait_for_pipe_off(dev_priv->dev, pipe);
+
+	/* Dont wait for pipe off, incase of CHT DSI CMD mode */
+	if (!cht_dsi_cmd_mode)
+		intel_wait_for_pipe_off(dev_priv->dev, pipe);
 }
 
 /*
-- 
1.9.1

